src/middlewared/middlewared/plugins/pool.py

```diff
@@ -4182,6 +4182,147 @@ def __handle_zfs_set_property_error(self, e, properties_definitions):
+            if len(m) != 3:
+                self.logger.debug('Invalid %s quota: %s', type.lower(), quota)
+                continue
+
+            entry = {
+                'type': type,
+                'id': int(m[0]),
+                'name': None,
+                'quota': int(m[2]),
+                'used_bytes': int(m[1]),
+                'used_percent': 0,
+            }
+            if entry['quota'] > 0:
+                entry['used_percent'] = entry['used_bytes']/entry['quota'] * 100
+
+            try:
+                if type == 'USER':
+                    entry['name'] = (
+                        await self.middleware.call('user.get_user_obj',
+                                                   {'uid': entry['id']})
+                    )['pw_name']
+                else:
+                    entry['name'] = (
+                        await self.middleware.call('group.get_group_obj',
+                                                   {'gid': entry['id']})
+                    )['gr_name']
+
+            except Exception:
+                self.logger.debug('Unable to resolve %s id %d to name',
+                                  type.lower(), entry['id'])
+                pass
+
+            quota_list.append(entry)
+
+        return filter_list(quota_list, filters, options)
+
+    @accepts(
+        Str('id', required=True),
+        List('quotas', items=[
+            Dict(
+                'quota_entry',
+                Str('type', enum=['USER', 'GROUP'], required=True),
+                Str('id', required=True),
+                Int('quota', required=True),
+            )
+        ], default=[{
+            'type': 'USER',
+            'id': '0',
+            'quota': 0
+        }])
+    )
+    @item_method
+    async def set_userquota(self, id, data):
+        """
+        Set a user or group quota on the specified ZFS dataset. The quota limits
+        the amount of disk space consumed by files that are owned by the specified
+        users or groups.
+
+        `id` the name of the target ZFS dataset.
+
+        `quotas` specifies a list of `quota_entry` entries to apply to dataset.
+
+        `quota_entry` entries have these required parameters:
+
+        `type`: specifies whether the quota applies to a user or a group.
+
+        `id`: the uid, gid, or name to which the quota applies.
+
+        `quota`: the quota size in bytes. Setting a value of `0` removes
+        the user or group quota.
+        """
+        MAX_QUOTAS = 100
+        dataset = (await self._get_instance(id))['name']
+        verrors = ValidationErrors()
+        if len(data) > MAX_QUOTAS:
+            verrors.add(
+                'pool.dataset.set_userquota.',
+                f'The number of userspace quotas that can be set in single API call is limited to {MAX_QUOTAS}.'
+            )
+
+        cmd = ['zfs', 'set']
+        for q in data:
+            type = q["type"].lower()
+            if not q["id"].isdigit():
```

Comment:
Should we perhaps verify digits as well if they exist or not ?

Reply:
Done